File: services/product-service/app/schemas/category.py
from pydantic import BaseModel, Field
from typing import Optional
from datetime import datetime

class CategoryBase(BaseModel):
    name: str = Field(..., min_length=1, max_length=255)
    slug: str = Field(..., min_length=1, max_length=255)
    description: Optional[str] = None
    parent_id: Optional[int] = None
    order: int = Field(default=0, ge=0)

class CategoryCreate(CategoryBase):
    pass

class CategoryUpdate(BaseModel):
    name: Optional[str] = Field(None, min_length=1, max_length=255)
    slug: Optional[str] = Field(None, min_length=1, max_length=255)
    description: Optional[str] = None
    parent_id: Optional[int] = None
    order: Optional[int] = Field(None, ge=0)

class CategoryResponse(CategoryBase):
    id: int
    created_at: datetime
    updated_at: Optional[datetime] = None

    class Config:
        from_attributes = True

File: shared/types/src/category.ts
import { z } from 'zod';

export const CategoryBaseSchema = z.object({
  name: z.string().min(1).max(255),
  slug: z.string().min(1).max(255),
  description: z.string().optional(),
  imageUrl: z.string().url().optional(),
  parentId: z.number().int().positive().optional(),
  sortOrder: z.number().int().min(0).default(0),
  metaTitle: z.string().max(255).optional(),
  metaDescription: z.string().optional(),
});

export const CategoryCreateSchema = CategoryBaseSchema;

export const CategoryUpdateSchema = CategoryBaseSchema.partial().extend({
  isActive: z.boolean().optional(),
});

export const CategoryResponseSchema = CategoryBaseSchema.extend({
  id: z.number().int().positive(),
  isActive: z.boolean(),
  createdAt: z.date(),
  updatedAt: z.date(),
  children: z.array(z.lazy(() => CategoryResponseSchema)).optional(),
});

export type CategoryBase = z.infer<typeof CategoryBaseSchema>;
export type CategoryCreate = z.infer<typeof CategoryCreateSchema>;
export type CategoryUpdate = z.infer<typeof CategoryUpdateSchema>;
export type CategoryResponse = z.infer<typeof CategoryResponseSchema>;
